### ISSUE TITLE
i18n

### ISSUE THREAD

**Issue by mehrdad-shokri** [react-boilerplate/react-boilerplate#2610]
**Is your feature request related to a problem? Please describe.**
Hi, This is not really a feature request rather a confusion in the docs. I've already reviewd [react boilerplate i18n docs](https://github.com/react-boilerplate/react-boilerplate/blob/master/docs/js/i18n.md) and [react intl docs](https://github.com/yahoo/react-intl/wiki/API) But can't find a solution.  

**Describe the solution you'd like**
I've added another locale called ``fa`` via ``npm run generate``.
Now some files are changed but I don't know where to put translations for another locale.  
Every ``messages.js`` file seems to correlate with ``en`` one.
Now I have 2 questions:  
First: How can I change my app in a way that ``messages.js`` relate to ``fa`` instead of ``en``?(Changing ``DEFAULT_LOCALE`` didn't fix it)
Two: Where should I define new locale translations?  
 
**Describe alternatives you've considered**
Right now I modify ``fa.json`` manually to add a translation string. But I think this is anti-pattern since it gets overwritten every time translations are compiled.

**Additional context**
Thanks for your awesome library and to the earth docs. It saves hours and hours of configuring tools and getting overwhelmed by the JS ecosystem!

**Comment by julienben** [comment]
I've never used the i18n setup but my understanding is that `messages.js` is where you "declare" all your messages with their path and their default value (`defaultMessage`).

So if you have `fa` as your `DEFAULT_LOCALE`, you should put all the `fa` text inside `messages.js`. I think that you still need `translations/fa.json` though.

**Comment by ghulamhussain** [comment]
@julienben  is rightly said. .messages.js is where you declare all your messages with their path and default value. but if you have fa as your default locale there is no need to put fa text in messages.js . the react intl will first try too look into fa.json if the path is missing there it will fallback to messages.js. it will always first look into json files.

**Comment by julienben** [comment]
@mehi-shokri Does this answer your question? Can you point out how/where the docs confused you? That way we can open a PR and fix it.

**Comment by mehrdad-shokri** [comment]
@julienben I'm sorry for late response, I will clarify more shortly.

**Comment by mehrdad-shokri** [comment]
Thank you guys for the answers. Please correct me if I'm wrong here.  
``react-intl`` first tries to read the json file correlated with the locale if it doesn't find corresponding string it fallbacks to ``messages.js``.  
Now I have 2 problems with this approach:  
First of all my ``.json`` translation files are compiled till I explicitly run ``npm run extract-intl`` then it becomes available for use (I can fill for example ``en.json`` which isn't the default locale).
Also, I'm using ``intl.formatMessage`` heavily in my app. `` <FormattedMessage {...messages.header} />`` fallbacks to default values defined in ``messages.json`` but ``intl.formatMessage({id: 'scopeName.header'}) will output the ``scopeName.header`` until I run a ``extract-intl`` in console.  
What's the workaround for this? since it's not very convenient to run ``extract-intl`` each time I add a message.  Any watch mode for this?

**Comment by Mensae** [comment]
@mehrdad-shokri 

**TL;DR** In `messages.js` you 'declare' your text in your *default* language. In any `translations/xx.json` you add your translations.

When you specify a default language, or set it otherwise, and it cannot find the correct `xx.json` file or the specific field is empty it will use `messages.js`. In `messages.js` you will specify the `defaultMessage` in your own default language, 'fa' in your case. You run `npm run extract-intl` usually once, when your component is ready. When you are still developing it will use `messages.js`.

The `npm run generate` script is there to give you something to start with. It uses a set of templates in order to create a new `container`, `component` or to add a language. Because of this the only other solution, for not to have the English language at all, would be to have templates for every language, but that is not very practical.
As a suggestion, you could, instead of using the `npm generate` script, add your own *templates* or *snippets* to the editor/IDE you are using and use these to add a new component/container.

In general, and for most users, I guess having English as default is the prefered option as this is most widely spoken. It depends of course on your intentded users and the area of the world you live in.

Besides all that, the proper way to use `intl.formatMessage()` is `intl.formatMessage(messages.header)`, where `messages.js` is like:

```js
export default defineMessages({
  header: {
    id: `${scope}.header`,
    defaultMessage: 'This is the Test component!',
  },
});
```

As now you're creating a new object, `{id: 'scopeName.header'}`, with no relation to `messages.js`.
I would however advice to avoid using the, lower level, `intl.formatMessage()` wherever possible.

Now with regard to your initial question, there is no direct correlation between any `messages.js` and  any `xx.json` file. Anything described in any `message.js` is indeed used as fallback when it cannot find either the default `xx.json` or the required entry within the default `xx.json`. `messages.js` is also used by the `npm run extract-intl` script to add the string in defaultValue to the specific field in every defined `xx.json` file, it will not overwrite any existing value though.
If you used `npm run generate` to add a new language, then a new `xx.json` is created. It will have all fields as already defined in any `messages.js`. These fields will be empty and there you need to add your translation for that specific language. If however you run `npm run extract-intl` any empty field will be filled with the default language, again, any existing value will not be overridden. I find this helpfull for translating, this instead of switching between files.

I hope it helps.

**Comment by mehrdad-shokri** [comment]
@Mensae thank you very much for the detailed response that was very clarifying.
